src/compiler/checker.ts

```diff
@@ -6052,6 +6056,32 @@ namespace ts {
                 }
                 return Ternary.True;
             }
+
+            function constructorRelatedTo(sourceSignature: Signature, targetSignature: Signature, reportErrors: boolean) {
+                if (sourceSignature && targetSignature && sourceSignature.declaration && targetSignature.declaration) {
+                    const sourceAccessibility = sourceSignature.declaration.flags & (NodeFlags.Private | NodeFlags.Protected);
+                    const targetAccessibility = targetSignature.declaration.flags & (NodeFlags.Private | NodeFlags.Protected);
+
+                    const isRelated = sourceAccessibility === targetAccessibility;
+                    if (!isRelated && reportErrors) {
+                        reportError(Diagnostics.Cannot_assign_a_0_constructor_type_to_a_1_constructor_type, flagsToString(sourceAccessibility), flagsToString(targetAccessibility));
+                    }
+
+                    return isRelated;
+                }
+
+                return true;
+
+                function flagsToString(flags: NodeFlags) {

```

Comment:
Move this outside the function and rename (`visibilityToString` ?) (most runtimes allocate a new closure per nested function on invocation, which is bad). You can also try factoring out the logic in lines ~16190-16199 to use this new function if you like.